README.md
# todolistUi

todolistUi

## Requirement

- install yarn
- install node (v12+)

## Testing and run

```zsh
$ yarn

// development
$ yarn run dev

// production
$ yarn run production

// run type check
$ yarn run type-check

// use eslint and prettier to format code
$ yarn run lint
```

open localhost:5000

## Docker

```zsh
// build images and start container in one line
docker-compose up -d --build

// go inside container
docker exec -it <containerId> /bin/bash

// check container logs
docker logs <containerId>

// remove and stop container
docker-compose down
```

open localhost:5000

## Contributing

Please refer to [CONTRIBUTING.md](https://github.com/yeukfei02/todolistUi/blob/master/CONTRIBUTING.md)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [0.2.10](https://github.com/yeukfei02/todolistUi/compare/v0.2.9...v0.2.10) (2021-07-25)


### Bug Fixes

* 🐛 remove not used file ([1505a1d](https://github.com/yeukfei02/todolistUi/commit/1505a1d2abcf1da1768a8266f6d15e8f6a1c5cca))

### [0.2.9](https://github.com/yeukfei02/todolistUi/compare/v0.2.8...v0.2.9) (2021-03-29)


### Bug Fixes

* 🐛 fix package.json ([722b040](https://github.com/yeukfei02/todolistUi/commit/722b04059ef915c8529a3bc7b286de56902e2289))

### [0.2.8](https://github.com/yeukfei02/todolistUi/compare/v0.2.7...v0.2.8) (2020-10-18)

### [0.2.7](https://github.com/yeukfei02/todolistUi/compare/v0.2.6...v0.2.7) (2020-10-17)

### [0.2.6](https://github.com/yeukfei02/todolistUi/compare/v0.2.5...v0.2.6) (2020-09-23)


### Features

* 🎸 add standard-version lib ([aaac3a4](https://github.com/yeukfei02/todolistUi/commit/aaac3a4629d7b39f08919c243e54d38690423079))

### [0.2.4](https://github.com/yeukfei02/todolistUi/compare/v0.2.3...v0.2.4) (2020-09-19)

### [0.2.3](https://github.com/yeukfei02/todolistUi/compare/v0.2.2...v0.2.3) (2020-09-10)

### [0.2.2](https://github.com/yeukfei02/todolistUi/compare/v0.2.1...v0.2.2) (2020-09-10)

### [0.2.1](https://github.com/yeukfei02/todolistUi/compare/v0.2.0...v0.2.1) (2020-09-09)

## [0.2.0](https://github.com/yeukfei02/todolistUi/compare/v0.1.9...v0.2.0) (2020-09-09)

### [0.1.9](https://github.com/yeukfei02/todolistUi/compare/v0.1.8...v0.1.9) (2020-06-27)

### [0.1.8](https://github.com/yeukfei02/todolistUi/compare/v0.1.7...v0.1.8) (2020-06-27)

### [0.1.7](https://github.com/yeukfei02/todolistUi/compare/v0.1.6...v0.1.7) (2020-06-07)

### [0.1.6](https://github.com/yeukfei02/todolistUi/compare/v0.1.5...v0.1.6) (2020-06-07)

### [0.1.5](https://github.com/yeukfei02/todolistUi/compare/v0.1.4...v0.1.5) (2020-06-07)

### [0.1.4](https://github.com/yeukfei02/todolistUi/compare/v0.1.3...v0.1.4) (2020-06-06)

### [0.1.3](https://github.com/yeukfei02/todolistUi/compare/v0.1.2...v0.1.3) (2020-06-06)

### [0.1.2](https://github.com/yeukfei02/todolistUi/compare/v0.1.1...v0.1.2) (2020-06-06)

### [0.1.1](https://github.com/yeukfei02/todolistUi/compare/v0.1.0...v0.1.1) (2020-06-06)

## 0.1.0 (2020-06-06)

### 0.2.5 (2020-09-23)


.eslintrc.js
module.exports = {
    parser: "@typescript-eslint/parser",
    extends: [
        "plugin:react/recommended", // Uses the recommended rules from @eslint-plugin-react
        "plugin:@typescript-eslint/recommended", // Uses the recommended rules from @typescript-eslint/eslint-plugin
        "prettier/@typescript-eslint", // Uses eslint-config-prettier to disable ESLint rules from @typescript-eslint/eslint-plugin that would conflict with prettier
        "plugin:prettier/recommended" // Enables eslint-plugin-prettier and eslint-config-prettier. This will display prettier errors as ESLint errors. Make sure this is always the last configuration in the extends array.
    ],
    parserOptions: {
        ecmaVersion: 2018, // Allows for the parsing of modern ECMAScript features
        sourceType: "module", // Allows for the use of imports
        ecmaFeatures: {
        jsx: true // Allows for the parsing of JSX
        }
    },
    rules: {
        // Place to specify ESLint rules. Can be used to overwrite rules specified from the extended configs
        // e.g. "@typescript-eslint/explicit-function-return-type": "off",
    },
    settings: {
        react: {
        version: "detect" // Tells eslint-plugin-react to automatically detect the version of React to use
        }
    }
};

pages/api/task/[taskParam].ts
import { NextApiRequest, NextApiResponse } from 'next';
import axios from 'axios';

const ROOT_URL = `https://todo-list-api-v1.herokuapp.com/api`;

export default async (req: NextApiRequest, res: NextApiResponse) => {
  console.log('req.query.taskParam = ', req.query.taskParam);
  const type = req.query.taskParam[0];
  const id = req.query.taskParam[1];
  const data = {
    taskMessage: req.query.taskParam[2],
    userId: req.query.taskParam[3],
  };

  switch (type) {
    case 'create-task':
      createTask(res, data);
      break;
    case 'getTaskById':
      getTaskById(res, id);
      break;
    case 'updateTaskById':
      updateTaskById(res, id, data);
      break;
    case 'deleteTaskById':
      deleteTaskById(res, id);
      break;
  }
};

async function createTask(res: NextApiResponse, data: any) {
  try {
    const result = await axios.post(`${ROOT_URL}/task/create-task`, data);

    res.status(200).json({
      result: result.data,
    });
  } catch (e) {
    res.status(400).json({
      message: e.message,
    });
  }
}

async function getTaskById(res: NextApiResponse, id: string) {
  try {
    const result = await axios.get(`${ROOT_URL}/task/${id}`);

    res.status(200).json({
      result: result.data,
    });
  } catch (e) {
    res.status(400).json({
      message: e.message,
    });
  }
}

async function updateTaskById(res: NextApiResponse, id: string, data: any) {
  try {
    const result = await axios.put(`${ROOT_URL}/task/${id}`, data);

    res.status(200).json({
      result: result.data,
    });
  } catch (e) {
    res.status(400).json({
      message: e.message,
    });
  }
}

async function deleteTaskById(res: NextApiResponse, id: string) {
  try {
    const result = await axios.delete(`${ROOT_URL}/task/${id}`);

    res.status(200).json({
      result: result.data,
    });
  } catch (e) {
    res.status(400).json({
      message: e.message,
    });
  }
}


pages/api/user/[...userParam].ts
import { NextApiRequest, NextApiResponse } from 'next';
import axios from 'axios';

const ROOT_URL = `https://todo-list-api-v1.herokuapp.com/api`;

export default async (req: NextApiRequest, res: NextApiResponse): Promise<any> => {
  const type = req.query.userParam[0];
  const username = req.query.userParam[1];

  switch (type) {
    case 'getMessage':
      getMessage(res, username);
      break;
    case 'createUser':
      createUser(res, username);
      break;
  }
};

async function getMessage(res: NextApiResponse, username: string) {
  try {
    const result = await axios.get(`${ROOT_URL}/user/get-user-id`, {
      params: {
        username: username,
      },
    });

    res.status(200).json({
      result: result.data,
    });
  } catch (e) {
    res.status(400).json({
      message: e.message,
    });
  }
}

async function createUser(res: NextApiResponse, username: string) {
  try {
    const result = await axios.post(`${ROOT_URL}/user/create-user`, {
      username: username,
    });

    res.status(200).json({
      result: result.data,
    });
  } catch (e) {
    res.status(400).json({
      message: e.message,
    });
  }
}


pages/api/task/[...taskParam].ts
import { NextApiRequest, NextApiResponse } from 'next';
import axios from 'axios';

const ROOT_URL = `https://todo-list-api-v1.herokuapp.com/api`;

export default async (req: NextApiRequest, res: NextApiResponse): Promise<any> => {
  const type = req.query.taskParam[0];
  const id = req.query.taskParam[1];

  switch (type) {
    case 'createTask':
      const createTaskData = {
        taskMessage: req.query.taskParam[2],
        userId: parseInt(req.query.taskParam[3], 10),
      };
      createTask(res, createTaskData);
      break;
    case 'getTaskByUserId':
      const userId = req.query.taskParam[2];
      getTaskByUserId(res, userId);
      break;
    case 'getTaskById':
      getTaskById(res, id);
      break;
    case 'updateTaskById':
      const updateTaskByIdData = {
        taskMessage: req.query.taskParam[2],
        userId: parseInt(req.query.taskParam[3], 10),
      };
      updateTaskById(res, id, updateTaskByIdData);
      break;
    case 'deleteTaskById':
      deleteTaskById(res, id);
      break;
  }
};

async function createTask(res: NextApiResponse, data: any) {
  try {
    const result = await axios.post(`${ROOT_URL}/task/create-task`, data);

    res.status(200).json({
      result: result.data,
    });
  } catch (e) {
    res.status(400).json({
      message: e.message,
    });
  }
}

async function getTaskByUserId(res: NextApiResponse, userId: string) {
  try {
    const result = await axios.get(`${ROOT_URL}/task`, {
      params: {
        userId: userId,
      },
    });

    res.status(200).json({
      result: result.data,
    });
  } catch (e) {
    res.status(400).json({
      message: e.message,
    });
  }
}

async function getTaskById(res: NextApiResponse, id: string) {
  try {
    const result = await axios.get(`${ROOT_URL}/task/${id}`);

    res.status(200).json({
      result: result.data,
    });
  } catch (e) {
    res.status(400).json({
      message: e.message,
    });
  }
}

async function updateTaskById(res: NextApiResponse, id: string, data: any) {
  try {
    const result = await axios.put(`${ROOT_URL}/task/${id}`, data);

    res.status(200).json({
      result: result.data,
    });
  } catch (e) {
    res.status(400).json({
      message: e.message,
    });
  }
}

async function deleteTaskById(res: NextApiResponse, id: string) {
  try {
    const result = await axios.delete(`${ROOT_URL}/task/${id}`);

    res.status(200).json({
      result: result.data,
    });
  } catch (e) {
    res.status(400).json({
      message: e.message,
    });
  }
}
